network-ntt/src/server.rs

```diff
@@ -0,0 +1,313 @@
+pub fn run<N: 'static>(config: Config, node: N)
+where
+    N: std::marker::Sync + std::marker::Send + server::Node + Clone + Copy,
+    <<N as network_core::server::Node>::BlockService as network_core::server::block::BlockService>::TipFuture: std::marker::Send,
+    <<N as network_core::server::Node>::BlockService as network_core::server::block::BlockService>::GetHeadersFuture: std::marker::Send,
+    <<N as network_core::server::Node>::BlockService as network_core::server::block::BlockService>::StreamBlocksToTipFuture: std::marker::Send,
+    <<N as network_core::server::Node>::TransactionService as network_core::server::transaction::TransactionService>::ProposeTransactionsFuture: std::marker::Send,
+{
+
+        let listener =
+            stream::iter_ok(config.listen)
+                .for_each(move |sockaddr| {
+                    let server = Server {
+                          node: node.clone(),
+                    };
+                    run_listen_socket(sockaddr, server) 
+                });
+        let connections =
+            stream::iter_ok(config.connect)
+                .for_each(move |sockaddr| {
+                let server = Server {
+                      node: node.clone(),
+                };
+                run_connect_socket(sockaddr, server)
+            });
+        tokio::run(connections.join(listener).map(|_| ()));
+}
+
+
+/// Run a server that will listen on a specific sockets
+/// and accept all incomming connections.
+/// Server maintains all of the incomming connection and
+/// `run_connection` is spawned on each of those connections.
+pub fn run_listen_socket<N: 'static>(sockaddr: SocketAddr, node: Server<N>) -> tokio::executor::Spawn
+where
+    N: server::Node + Clone,
+    N: std::marker::Send,
+    <<N as network_core::server::Node>::BlockService as network_core::server::block::BlockService>::TipFuture: std::marker::Send,
+    <<N as network_core::server::Node>::BlockService as network_core::server::block::BlockService>::GetHeadersFuture: std::marker::Send,
+    <<N as network_core::server::Node>::BlockService as network_core::server::block::BlockService>::StreamBlocksToTipFuture: std::marker::Send,
+    <<N as network_core::server::Node>::TransactionService as network_core::server::transaction::TransactionService>::ProposeTransactionsFuture: std::marker::Send,
+{
+    let server = TcpListener::bind(&sockaddr)
+        .expect("Can't bind socket")
+        .incoming()
+        .map_err(move |_err| ())
+        .for_each(move |stream| {
+            let node = node.clone();
+            Connection::accept(stream)
+                .map_err(move |_err| unimplemented!("[2]"))
+                .and_then(move |connection| {
+                    tokio::spawn(run_connection(node, connection))
+                })
+        });
+    tokio::spawn(server)
+}
+
+/// Connect to another client.
+/// `run_connection` is spawned on the single heavyweight
+/// connection.
+pub fn run_connect_socket<N: 'static>(
```

Comment:
I would put this away for another PR, when we design some sort of client API abstraction.

Reply:
I'd keep it here due to the nature of the n-t, the client in network transport implementation will be some other thing.